src/engine/renderer/gl_shader.cpp

```diff
@@ -55,27 +63,236 @@ GLShader_motionblur                      *gl_motionblurShader = nullptr;
+			if (!foundCommentEnd)
+			{
+				// If we don't find the end of a comment:
+				// If it' a block comment we failed to close then keep everything
+				// before it and after. Before it is code and keeping what's
+				// after makes it easier to diagnose what's wrong.
+				// If we can't close a line comment (i.e. no new line), keep everything
+				// before it as that's code, but forget the line comment and what follows.
+				// Either way we are done.
+				if (commentType == CommentType::Block)
+				{
+					size_t remainingLength = source.length() - sourcePos;
+					keep(remainingLength);
+				}
+				else
+				{
+					size_t codeLength = commentStartPos - sourcePos;
+					keep(codeLength);
+				}
+				break;
+			}
+			// We found a start and end comment. keep the what's before the comment.
+			// as that's code. Then continue after the comment.
+			// Note that the new line that exists after comments is kept to avoid
+			// splicing any lines together without gaps that would create new unexpected tokens.
+			size_t codeLength = commentStartPos - sourcePos;
+			keep(codeLength);
+			sourcePos = commentEndPos;
+		}
+		source.resize(keepPos);
+	}
+
+	// Somewhat emulate what buildshader.sh does or doesn't do.
+	// Not wildly efficient but doesn't have to be as it's currently
+	// only used as part of a debugging only feature on small shader files.
+	void NormalizeShaderText( std::string& text )
 	{
-		delete _shaders[ i ];
+		size_t pos = 0;
+		while ((pos = text.find("\r\n", pos)) != std::string::npos)
+			text.erase(text.begin()+pos);
+		StripComments(text);
 	}
 
-	_shaders.erase( _shaders.begin(), _shaders.end() );
+	std::string GetShaderFilename(Str::StringRef filename)
+	{
+		std::string shaderBase = GetShaderPath();
+		if (shaderBase.empty())
+			return shaderBase;
+		std::string shaderFileName = FS::Path::Build(shaderBase, filename);
+		return shaderFileName;
+	}
 
-	for ( std::size_t i = 0; i < _deformShaders.size(); i++ )
+	std::string GetShaderText(Str::StringRef filename)
 	{
-		glDeleteShader( _deformShaders[ i ] );
+		Log::Debug("loading shader '%s'\n", filename);
+
+		// Shader type should be set during initialisation.
+
+		if (shaderKind == ShaderKind::BuiltIn)
+		{
+			// Look for the shader internally. If not found, look for it externally.
+			// If found neither internally or externally of if empty, then Error.
+			auto text_ptr = GetInternalShader(filename);
+			if (text_ptr == nullptr)
+				ThrowShaderError(Str::Format("No shader found for shader: %s", filename));
+			return text_ptr;
+		}
+		else if (shaderKind == ShaderKind::External)
+		{
+			std::string shaderText;
+			std::string shaderFilename = GetShaderFilename(filename);
+			std::error_code openErr;
+
+			FS::File shaderFile = FS::RawPath::OpenRead(shaderFilename, openErr);
+			if (openErr)
+				ThrowShaderError(Str::Format("Cannot load shader from file: %s\n", shaderFilename));
```

Comment:
You want to print the actual error message as well, which you can obtain from openErr.message(). The same applies to all other std::error_code uses.